packages/kit/src/runtime/server/respond.js

```diff
@@ -283,11 +283,17 @@ export async function respond(request, options, manifest, state) {
 					prerender = node.prerender ?? prerender;
 				} else if (route.page) {
 					const nodes = await load_page_nodes(route.page, manifest);
-					config = get_page_config(nodes);
+					config = get_page_config(nodes) ?? config;
 					prerender = get_option(nodes, 'prerender') ?? false;
 				}
 
-				state.before_handle(event, config, prerender);
+				if (state.before_handle) {
+					state.before_handle(event, config, prerender);
+				}
+
+				if (state.emulator && !prerender) {
```

Comment:
If the platform gives you access to the production datastore then I think this would be overly restrictive. I think it really depends how the user has setup their `.env` file as to whether you'd want to let them access certain things. But I don't know if there's a way that we can tell if it's using `.env` vs `.env.local` or if people even consistently use the former for production and the latter locally

Reply:
`env` is whatever Vite says it is, so ambient env vars will take precedence, falling back to `.env.local` if it exists then `.env` if it doesn't (and the various other permutations involving `mode` etc).

But Vite doesn't know about `wrangler.toml`. I guess if it's set up correctly (with an `id` and a `preview_id`) then we could ensure that we use `preview_id` during dev and preview but `id` during prerendering? I'm not sure how we'd make that work with `getBindingsProxy` though, as I don't see an option for specifying dev vs prod — this would be a question for @dario-piotrowicz

Reply:
following conversation with @dario-piotrowicz and @IgorMinar: we can't use bindings during prerendering, because the data accessed via `getBindingsProxy` is local (it lives in a `.wrangler` directory). Using production data would be a whole other thing. It _might_ work in future, and other adapters might want to populate `platform` during prerendering, so we should make that work. But the `platform` function in the Cloudflare adapters should return `undefined`, or maybe something like this:

```js
if (prerender) {
  const env = {};

  for (const key of proxy.bindings) {
    Object.defineProperty(env, key, {
      get: () => {
        throw new Error(`Cannot access platform.env.${key} in a prerenderable route`);
      }
    });
  }

  return { env };
}

return { env: proxy.bindings };
```